Feature/migrate to kubebuilder v3

itamar-marom opened the pull request:
Fixes #24

### Description
Currently, the project is not following Kubebuilder's native structure. This customization can be a real pain for new developers who are familiar with Kubebuilder, and also make new API creation difficult.

This PR updates Kubebuilder to v3 including back to Kubebulder's native structure.  

<!-- Describe key changes made in the patch. -->
[Following the migration guide](https://book.kubebuilder.io/migration/legacy/migration_guide_v2tov3.html)  
This PR is something that has to be done before this project gains more users and becomes bigger. This project is a mutation of the Operator SDK and the Kubebuilder project.
In this MR I'm migrating the project to Kubebuilder while upgrading Kubebuilder to v3 in order to eliminate the pain of migrating at another time.

Something that is notable here is that autoscaling/v2alpha2 is now deprecated in new clusters (For example, update your kind cluster and try) - Therefore, the project is now using autoscaling/v2 HPAs.  

The deployment also had a big change since Kubebuilder deploys more resources that benefit us and help us to manage the controller much more easily.  

I will note the Kubebuilder is a Kubernetes community project and it is becoming the de-facto for building controllers - we must follow the guidelines.

Tested on:
- Kubernetes version -> v1.25
- Kubebuilder version -> 3.9.1
- Go version -> 1.20.2

<hr>

This PR has:
- [x] has been tested on a real K8S cluster to ensure creating a brand new Druid cluster works.
- [ ] been tested for backward compatibility on a real K*S cluster by applying the changes introduced here on an existing Druid cluster. If there are any backward incompatible changes then they have been noted in the PR description.
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious to an unfamiliar reader.
- [x] added documentation for new or modified features or behaviors.

<hr>

##### Key changed/added files in this PR
* `deploy` structure that changes to `config` and deploys more resources by Kubebuilder best practices.
* The `Makefile` file changed to support the changes.  
* Some code in `main.go` and the `Reconcile` function went through some little changes.
* docs updated
* helm chart has changed to mirror the `config` directory

AdheipSingh commented on go.mod:
```diff
@@ -3,13 +3,14 @@ module github.com/datainfrahq/druid-operator
 go 1.19
 
 require (
+	github.com/druid-io/druid-operator v0.0.0-20230223202126-9cec3f2814d3
```
not needed, the repo is deprecated.

itamar-marom replied:
Fixed :)

AdheipSingh reviewed (commented):
This will also need changes in the documentation, helm charts.

AdheipSingh commented:
Thanks for taking this @itamar-marom . Also to test your PR what k8s version are you testing on ? would be good to note and add it the compatibility matrix.

itamar-marom commented:
Hi, @AdheipSingh Thanks for already looking at the PR. It is still not ready (that's why the WIP) and I don't want you to waste your time :), I'll ping you when it's ready. I'll reply to your comments, thank you!

itamar-marom commented:
I'm currently dealing with this issue: https://github.com/kubernetes-sigs/kubebuilder/issues/2026 - this is why it takes more time.

AdheipSingh commented:
> I'm currently dealing with this issue: [kubernetes-sigs/kubebuilder#2026](https://github.com/kubernetes-sigs/kubebuilder/issues/2026) - this is why it takes more time.

Let me know if you hit any other blocker, happy to brain storm together.

itamar-marom commented:
I fixed it. Wrote about it [here](https://kubernetes.slack.com/archives/C04F4M6HT2L/p1678643030512149)  
I don't know why it makes a difference.  
Anyway, another thing is that tests are not working for me since my cluster support autoscaling/v2 and not v2beta2. I updated the code to supportt v2 but let's talk about it, and if you think this is a bad idea I'll revert back.

itamar-marom commented:
Doing some more testing, will update tomorrow

itamar-marom commented on main.go:
```diff
@@ -1,21 +1,42 @@
-	"github.com/datainfrahq/druid-operator/controllers/druid"
-	// +kubebuilder:scaffold:imports
+	druidcontrollers "github.com/druid-io/druid-operator/controllers/druid"
+	//+kubebuilder:scaffold:imports
```
Fixed :)

itamar-marom commented on main.go:
```diff
@@ -25,48 +46,66 @@ var (
+		Log:           ctrl.Log.WithName("controllers").WithName("Druid"),
+		ReconcileWait: druid.LookupReconcileTime(),
+		Recorder:      mgr.GetEventRecorderFor("druid-operator"),
+	}).SetupWithManager(mgr); err != nil {
```
Not sure it is necessary since `main.go` is the only place we're using it. But I added it back.

cyril-corbon commented on .github/workflows/docker-image.yml:
```diff
@@ -15,8 +15,8 @@ jobs:
-      run: make template
+      run: make helm-template
     - name: Run e2e tests
       run: make e2e
```
```suggestion
      run: make e2e

```

cyril-corbon commented on chart/templates/_helpers.tpl:
```diff
@@ -54,9 +54,5 @@ app.kubernetes.io/instance: {{ .Release.Name }}
-{{- if .Values.serviceAccount.create }}
 {{- default (include "druid-operator.fullname" .) .Values.serviceAccount.name }}
-{{- else }}
-{{- default "default" .Values.serviceAccount.name }}
-{{- end }}
 {{- end }}
```
I prefer to let `druid-operator.fullname`and remove this part

itamar-marom replied:
Fixed

cyril-corbon commented on chart/templates/deployment.yaml:
```diff
@@ -18,50 +24,79 @@ spec:
+            {{- toYaml .Values.securityContext | nindent 12 }}
+      securityContext:
+        {{- toYaml .Values.podSecurityContext | nindent 8 }}
+      serviceAccountName: {{ include "druid-operator.serviceAccountName" . }}
```
```suggestion
      serviceAccountName: {{ include "druid-operator.fullname" . }}
```

itamar-marom replied:
Fixed

cyril-corbon commented on chart/templates/rbac.yaml:
```diff
@@ -0,0 +1,283 @@
+- apiGroups:
+    - apps
+  resources:
+    - deployments
```
```suggestion
    - deployments
    - statefulsets
```

itamar-marom replied:
I can do that, but this YAML are generated bu Kubebuilder, so next time we will generate the YAMLs it will be like that again.

cyril-corbon replied:
kubebuilder now generate the helm chart too ? 👀

cyril-corbon replied:
if so just resolve all the comments on the helm chart, I'll check the doc too :D

itamar-marom replied:
Oh, I wish hehe. I just mean that he generates RBAC, and what I did was just copy-paste what he generated - that means that if I will do these changes, the next time that someone will change permissions, he will need to manually do these changes again.

itamar-marom replied:
@cyril-corbon we can talk about it on slack if you want it will be faster :)

AdheipSingh replied:
We might want to check the controller-gen tags on which generate rbac. 
here https://github.com/datainfrahq/druid-operator/blob/master/controllers/druid/druid_controller.go#L41

itamar-marom replied:
Yes, this is where all the manager's permissions are being generated from. We can removed whatever we want.

cyril-corbon commented on chart/templates/rbac.yaml:
```diff
@@ -0,0 +1,283 @@
+    - patch
+    - update
+    - watch
+- apiGroups:
+    - apps
+  resources:
+    - statefulsets
+  verbs:
+    - create
+    - delete
+    - get
+    - list
+    - patch
+    - update
+    - watch
```
```suggestion
```

AdheipSingh commented on main.go:
```diff
@@ -25,37 +45,50 @@ var (
 		LeaderElection:         enableLeaderElection,
-		LeaderElectionID:       "e6946145.apache.org",
-		Namespace:              os.Getenv("WATCH_NAMESPACE"),
-		NewCache:               watchNamespaceCache(),
```
Using ```NewCache: watchNamespaceCache()``` we can specify multiple namespace to watch or a single namespace. IMO it should be added, since default support is only for a single namespace ( unless the library changed internally, i will check this to confirm  )

itamar-marom replied:
Fixed

AdheipSingh commented on go.sum:
```diff
@@ -58,6 +58,8 @@ github.com/davecgh/go-spew v1.1.0/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSs
 github.com/davecgh/go-spew v1.1.1 h1:vj9j/u1bqnvCEfJOwUhtlOARqs3+rkHYY13jYWTU97c=
 github.com/davecgh/go-spew v1.1.1/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
 github.com/docopt/docopt-go v0.0.0-20180111231733-ee0de3bc6815/go.mod h1:WwZ+bS3ebgob9U8Nd0kOddGdZWjyMGR8Wziv+TBNwSE=
+github.com/druid-io/druid-operator v0.0.0-20230223202126-9cec3f2814d3 h1:Sd8295B7rOYBG4ISiZFgU+CW1XNyyMXYC50qODueVgQ=
+github.com/druid-io/druid-operator v0.0.0-20230223202126-9cec3f2814d3/go.mod h1:tbtTlEK57N7cwIF/zZeQBt+yYqdQ8NhtxMbgd6NedwA=
```
just wondering whats exactly adding these modules. this repo is deprecated.

itamar-marom replied:
Fixed, probably leftovers

cyril-corbon commented on chart/templates/rbac.yaml:
```diff
@@ -0,0 +1,283 @@
+  resources:
+    - events
+  verbs:
+    - create
+    - delete
+    - get
+    - list
+    - patch
+    - update
+    - watch
```
```suggestion
    - create
    - watch
```

itamar-marom replied:
Can you tell me which resource is that? Anyway changes to permissions are made from Kubebuilder markers like you did below, you can also send this as a change suggestion 😄

cyril-corbon commented on chart/templates/rbac.yaml:
```diff
@@ -0,0 +1,283 @@
+- apiGroups:
+    - ""
+  resources:
+    - configmaps
```
could we group theses resources by apiGroups ? 
like 
```
- configmaps
- persistentvolumeclaims
- services
- pods
```

cyril-corbon commented on chart/templates/rbac.yaml:
```diff
@@ -0,0 +1,283 @@
+  name: {{ include "druid-operator.fullname" . }}-leader-election-role
+subjects:
+  - kind: ServiceAccount
+    name: {{ include "druid-operator.serviceAccountName" . }}
```
```suggestion
    name: {{ include "druid-operator.fullname" . }}
```

itamar-marom replied:
Fixed all SA name in all places

cyril-corbon commented on chart/templates/rbac.yaml:
```diff
@@ -0,0 +1,283 @@
+  name: {{ include "druid-operator.fullname" . }}-manager-role
+subjects:
+  - kind: ServiceAccount
+    name: {{ include "druid-operator.serviceAccountName" . }}
```
```suggestion
    name: {{ include "druid-operator.fullname" . }}
```

itamar-marom replied:
Fixes

cyril-corbon commented on chart/templates/rbac.yaml:
```diff
@@ -0,0 +1,283 @@
+  name: {{ include "druid-operator.fullname" . }}-proxy-role
+subjects:
+  - kind: ServiceAccount
+    name: {{ include "druid-operator.serviceAccountName" . }}
```
```suggestion
    name: {{ include "druid-operator.fullname" . }}
```

itamar-marom replied:
Fixed

cyril-corbon commented on chart/templates/service.yaml:
```diff
@@ -0,0 +1,20 @@
+      protocol: TCP
+      targetPort: https
+  selector:
+    control-plane: controller-manager
```
```suggestion
    control-plane: controller-manager

```

cyril-corbon commented on controllers/druid/druid_controller.go:
```diff
@@ -40,16 +40,25 @@ func NewDruidReconciler(mgr ctrl.Manager) *DruidReconciler {
+// +kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=persistentvolumeclaims,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=events,verbs=get;list;watch;create;update;patch;delete
```
```suggestion
// +kubebuilder:rbac:groups="",resources=events,verbs=get;list;watch;create;
```

AdheipSingh commented on chart/templates/deployment.yaml:
```diff
@@ -18,50 +24,79 @@ spec:
           readinessProbe:
             {{ toYaml . | nindent 12 }}
           {{- end }}
-          env:
-            {{- range $key, $value :=  .Values.env }}
-            - name: {{ $key }}
-              value: {{ tpl $value $ | quote }}
-            {{- end }}
```
why did we remove this ? 
In values.yaml there are eny getting referenced.

itamar-marom replied:
It is still here just moved a line. look at line 73

AdheipSingh replied:
ok sorry missed it

AdheipSingh commented on chart/templates/deployment.yaml:
```diff
@@ -18,50 +24,79 @@ spec:
           resources:
             {{- toYaml .Values.resources | nindent 12 }}
-      {{- with .Values.nodeSelector }}
-      nodeSelector:
-        {{- toYaml . | nindent 8 }}
-      {{- end }}
-      {{- with .Values.affinity }}
-      affinity:
-        {{- toYaml . | nindent 8 }}
-      {{- end }}
-      {{- with .Values.tolerations }}
-      tolerations:
-        {{- toYaml . | nindent 8 }}
-      {{- end }}
```
same, any specific reason to remove this ?

itamar-marom replied:
Also just moved lines, look at line 29

AdheipSingh commented on chart/templates/role.yaml:
```diff
@@ -1,204 +0,0 @@
-{{- if .Values.rbac.create }}
-{{- $env := .Values.env }}
-{{- if and ($env.WATCH_NAMESPACE) (ne $env.WATCH_NAMESPACE "default") }}
-# Split WATCH_NAMESPACE by commas and loop on them
-{{- $watchedNamespaces := (split "," $env.WATCH_NAMESPACE) -}}
-{{- range $watchedNamespaces }}
----
```
@itamar-marom there is a conditional check, which is needed to scope the operator. Operator can watch multiple namespace, single namespace or all namespaces. On this either a role or clusterrole is created.

itamar-marom replied:
OK, I understand. Fixed it but change it a little bit so there will be not extra role `druid-operator-manager-role-sc` - In the case of WATCH_NAMESPACES, the default ClusterRole will be created with only storageclass permissions.

itamar-marom replied:
Generally, the role permissions are the same, also naming wise if you look at the chart, and the role.yaml under the `deploy` directory.
</br>
This is a bit tricky since it really depends on how the user deployed the operator. In the case of helm, it will be responsible to change the rbac to the correct one. In the case of regular deployment, the user should notice that the names are equal between the old and the new deployment.

itamar-marom replied:
I detailed more in the migration docs. Eventually, it is important to check that the naming of the objects is the same in both deployments.

AdheipSingh commented on docs/kubebuilder_v3_migration.md:
```diff
@@ -0,0 +1,107 @@
+- New `labelSelector` value - `Deployment` cannot be updated in place.
+
+This guide will help you go through the migration to Kubebuilder V3.</br>
+<b>Note: These guides assumes that the current operator is running in the `druid-operaotr` namespace.</b>
```
nit ```druid-operaotr``` to ```druid-operator```

itamar-marom replied:
Fixed

AdheipSingh commented:
cc @cintoSunny @harinirajendran  if you can help in reviewing this too.

AdheipSingh commented on Makefile:
```diff
@@ -52,8 +51,7 @@ help: ## Display this help.
 manifests: controller-gen ## Generate WebhookConfiguration, ClusterRole and CustomResourceDefinition objects.
-	$(CONTROLLER_GEN) crd:generateEmbeddedObjectMeta=true paths="./..." output:crd:artifacts:config=deploy/crds
-	$(CONTROLLER_GEN) crd:generateEmbeddedObjectMeta=true paths="./..." output:crd:artifacts:config=chart/templates/crds/
+	$(CONTROLLER_GEN) rbac:roleName=manager-role crd:maxDescLen=0 webhook paths="./..." output:crd:artifacts:config=config/crd/bases
```
question: when i do ```make manifests``` the above command generate the CRD for both config/crd/bases and helm chart crds. How is it going to be with these changes ?

itamar-marom replied:
It will generate the entire config directory. It won't touch the chart at all so this will still be a manual step. But we need a solution to that. I don't think there is something like that out there.

itamar-marom replied:
This is Kubebuilder's default

AdheipSingh replied:
Yes, so to solve this i had added
```	$(CONTROLLER_GEN) crd:generateEmbeddedObjectMeta=true paths="./..." output:crd:artifacts:config=chart/templates/crds/```
So this way ```make manifests``` updates both the places. Lets keep it this way :)

AdheipSingh commented on Makefile:
```diff
@@ -69,32 +67,63 @@ vet: ## Run go vet against code.
 .PHONY: docker-build
 docker-build: test ## Build docker image with the manager.
-	docker build -t ${IMG}:${IMG_TAG} .
+	docker build -t ${IMG} .
```
why no tag ?

itamar-marom replied:
Fixed both `make manifests` & docker-build (and docker-push and docker-buildx)

AdheipSingh commented:
LGTM with the PR, ill do some final review and tests and wait if other want to comment or review.
